File: app/src/main/java/com/andb/apps/todo/AddTask.java
package com.andb.apps.todo;

import android.app.Activity;
import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.text.TextUtils;
import android.transition.TransitionManager;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.TimePicker;

import com.andb.apps.todo.filtering.FilteredLists;
import com.andb.apps.todo.filtering.Filters;
import com.andb.apps.todo.lists.TaskList;
import com.andb.apps.todo.lists.interfaces.TaskListInterface;
import com.andb.apps.todo.objects.Tasks;
import com.andb.apps.todo.settings.SettingsActivity;
import com.andrognito.flashbar.Flashbar;
import com.fastaccess.datetimepicker.callback.DatePickerCallback;
import com.fastaccess.datetimepicker.callback.TimePickerCallback;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.jaredrummler.cyanea.Cyanea;
import com.jaredrummler.cyanea.app.CyaneaAppCompatActivity;

import org.joda.time.DateTime;
import org.joda.time.LocalTime;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Collections;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;


public class AddTask extends CyaneaAppCompatActivity implements DatePickerCallback, TimePickerCallback {


    private int position = 2;
    private boolean full;
    private boolean editing;
    public static boolean taskEditingLoaded = false;
    public static boolean tagEditingLoaded;
    private int tagPosition = 0;
    private int taskPosition = 0;

    public ArrayList<Tasks> taskList = new ArrayList<>();

    public DateTime taskDateTime;
    long timeMin = 00;
    boolean timeHasBeenSet;


    private RecyclerView mRecyclerView;
    private static AddTaskAdapter mAdapter;
    private RecyclerView.LayoutManager mLayoutManager;
    private RecyclerView tagRecyclerView;
    private static AddTaskTagAdapter tagAdapter;
    private RecyclerView.LayoutManager tagLayoutManager;

    public ArrayList<String> itemsList = new ArrayList<>();

    static Activity activity;

    ImageView resetTimeButton;

    boolean notified;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_add_task);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        resetTimeButton = (ImageView) findViewById(R.id.resetTimeButton);
        final TextView timeText = (TextView) findViewById(R.id.dateTimeText);

        resetTimeButton.setVisibility(View.GONE);
        resetTimeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                taskDateTime = new DateTime(3000, 1, 1, 0, 0);
                resetTimeButton.setVisibility(View.GONE);
                timeText.setText(getResources().getText(R.string.add_time_text));
                timeHasBeenSet = false;
            }
        });

        Bundle bundle = getIntent().getExtras();
        taskEditingLoaded = false;
        tagEditingLoaded = false;
        if (bundle.containsKey("edit"))
            editing = bundle.getBoolean("edit");


        activity = this;
        timeHasBeenSet = false;


        if (editing) {


            boolean browse = bundle.getBoolean("browse");

            if (browse) {
                taskList = FilteredLists.browseTaskList;
            } else {
                taskList = FilteredLists.inboxTaskList;
            }

            taskPosition = bundle.getInt("editPos");
            Log.d("taskPosition", Integer.toString(taskPosition));

            notified = taskList.get(taskPosition).isNotified();


            //itemsList = taskList.get(position).getAllListItems();
            prepareForEditing(taskPosition);

            taskDateTime = taskList.get(taskPosition).getDateTime();


            if (!taskDateTime.equals(new DateTime(3000, 1, 1, 0, 0))) {
                timeText.setText(taskDateTime.toString("MMM d, h:mm a"));
                resetTimeButton.setVisibility(View.VISIBLE);
            }

        } else {
            taskDateTime = new DateTime(3000, 1, 1, 0, 0);
            prepareItems();
            prepareItemsRecyclerView(false);
            AddTaskTagAdapter.tagList.clear();
            if (!Filters.getCurrentFilter().isEmpty()) {
                AddTaskTagAdapter.tagList.addAll(Filters.getCurrentFilter());
            }
            prepareTagsRecyclerView(false);
            switchList();

        }


        checkAddListItem();

        checkAddTagItem();

        checkAddTime();

        fabAddList();


    }


    public void prepareItems() {
        for (int i = 0; i < position; i++) {
            itemsList.add("");
        }

    }

    public void prepareForEditing(int position) {
        boolean isItems = taskList.get(position).isListItems();
        boolean isTags = taskList.get(position).isListTags();
        EditText taskName = (EditText) findViewById(R.id.taskName);

        taskName.setText(taskList.get(position).getListName());

        //AddTaskAdapter.taskList.clear();
        AddTaskTagAdapter.tagList.clear();


        if (isItems) {
            itemsList = new ArrayList<>(taskList.get(position).getAllListItems());

            prepareItemsRecyclerView(true);

            //copy-pasted from switchList, starts checked
            final ViewGroup task_layout = (ViewGroup) findViewById(R.id.task_layout);

            Switch switch_task = (Switch) findViewById(R.id.switch_task);
            switch_task.setChecked(true);
            task_layout.setVisibility(View.VISIBLE);
            switch_task.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {


                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    if (isChecked) {
                        TransitionManager.beginDelayedTransition(task_layout);
                        task_layout.setVisibility(View.VISIBLE);

                    } else {
                        TransitionManager.beginDelayedTransition(task_layout);
                        task_layout.setVisibility(View.INVISIBLE);
                    }
                }
            });


        } else {
            prepareItemsRecyclerView(false);
            switchList();
        }


        if (isTags) {

            for (int i = 0; i < taskList.get(position).getListTagsSize(); i++) {
                AddTaskTagAdapter.tagList.add(taskList.get(position).getListTags(i));

            }

            prepareTagsRecyclerView(true);


        } else {
            prepareTagsRecyclerView(false);
            switchList();
        }


    }


    public void prepareItemsRecyclerView(boolean edit) {
        mRecyclerView = (RecyclerView) findViewById(R.id.itemRecyclerView);

        // use this setting to improve performance if you know that changes
        // in content do not change the layout size of the RecyclerView
        //mRecyclerView.setHasFixedSize(true);

        // use a linear layout manager
        mLayoutManager = new LinearLayoutManager(this);
        mRecyclerView.setLayoutManager(mLayoutManager);

        // specify an adapter (see also next example)
        mAdapter = new AddTaskAdapter(itemsList, edit, taskPosition);
        mRecyclerView.setAdapter(mAdapter);

        // Create an `ItemTouchHelper` and attach it to the `RecyclerView`
        ItemTouchHelper ith = new ItemTouchHelper(_ithCallback);
        ith.attachToRecyclerView(mRecyclerView);


    }

    public void prepareTagsRecyclerView(boolean edit) {


        tagRecyclerView = (RecyclerView) findViewById(R.id.tagRecyclerView);

        // use this setting to improve performance if you know that changes
        // in content do not change the layout size of the RecyclerView
        //tagRecyclerView.setHasFixedSize(true);

        // use a linear layout manager
        tagLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        tagRecyclerView.setLayoutManager(tagLayoutManager);

        // specify an adapter (see also next example)
        tagAdapter = new AddTaskTagAdapter(taskList, edit, taskPosition);
        tagRecyclerView.setAdapter(tagAdapter);
    }


    public void switchList() {

        final ViewGroup task_layout = (ViewGroup) findViewById(R.id.task_layout);

        final Switch switch_task = (Switch) findViewById(R.id.switch_task);

        switch_task.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {


            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    TransitionManager.beginDelayedTransition(task_layout);
                    task_layout.setVisibility(View.VISIBLE);

                } else {
                    TransitionManager.beginDelayedTransition(task_layout);
                    task_layout.setVisibility(View.GONE);

                }
            }
        });
    }


    public void checkAddListItem() {
        ConstraintLayout addTask = (ConstraintLayout) findViewById(R.id.addButton);

        addTask.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                position++;
                ;
                itemsList.add("");
                Log.d("addingValue", "adding #" + position);
                mRecyclerView.setItemViewCacheSize(itemsList.size());
                mAdapter.focused = true;
                mAdapter.notifyItemInserted(itemsList.size() - 1);
                mRecyclerView.smoothScrollToPosition(position);

                //mAdapter.notifyDataSetChanged();
            }
        });
    }

    public void checkAddTagItem() {
        ImageView addTag = (ImageView) findViewById(R.id.tagAddButton);
        addTag.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                tagPosition++;
                Intent selectTag = new Intent(AddTask.this, TagSelect.class);
                selectTag.putExtra("isTaskCreate", true);
                startActivity(selectTag);
            }
        });
    }

    public void checkAddTime() {
        ImageView timeButton = (ImageView) findViewById(R.id.timeButton);
        final ImageView dateButton = (ImageView) findViewById(R.id.dateButton);
        final TextView timeText = (TextView) findViewById(R.id.dateTimeText);


        final long timeMin = 00;

        dateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DatePickerDialog dialog = new DatePickerDialog(AddTask.this, new DatePickerDialog.OnDateSetListener() {
                    @Override
                    public void onDateSet(DatePicker datePicker, int i, int i1, int i2) {
                        TextView timeText = (TextView) findViewById(R.id.dateTimeText);

                        taskDateTime = taskDateTime.withDate(i, i1 + 1, i2);
                        if (!timeHasBeenSet) {
                            taskDateTime = taskDateTime.withTime(23, 59, 59, 0);
                            timeText.setText(taskDateTime.toString("MMM d"));

                        } else {
                            timeText.setText(taskDateTime.toString("MMM d, h:mm a"));
                        }
                        Log.d("dateTime", taskDateTime.toString());

                        if (editing) {
                            if (taskDateTime.isAfter(DateTime.now()))
                                notified = false;
                            else if (taskDateTime.isBefore(taskList.get(taskPosition).getDateTime()))
                                notified = taskList.get(taskPosition).isNotified();
                        }

                        resetTimeButton.setVisibility(View.VISIBLE);

                    }
                }, DateTime.now().getYear(), DateTime.now().getMonthOfYear() - 1, DateTime.now().getDayOfMonth());
                dialog.show();


//                DatePickerFragmentDialog.newInstance(DateTimeBuilder.newInstance()
//                        .withMinDate(Calendar.getInstance().getTimeInMillis()))
//                        .show(getSupportFragmentManager(), "DatePickerFragmentDialog");

            }
        });

        timeButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                TimePickerDialog timePickerDialog = new TimePickerDialog(AddTask.this, new TimePickerDialog.OnTimeSetListener() {
                    @Override
                    public void onTimeSet(TimePicker timePicker, int i, int i1) {
                        TextView timeText = (TextView) findViewById(R.id.dateTimeText);
                        timeHasBeenSet = true;
                        LocalTime localTime = new LocalTime().withHourOfDay(i).withMinuteOfHour(i1);
                        localTime = localTime.secondOfMinute().setCopy(0);
                        if (taskDateTime.isEqual(new DateTime(3000, 1, 1, 0, 0))) {
                            taskDateTime = new DateTime(DateTime.now().getYear(), DateTime.now().getMonthOfYear(), DateTime.now().getDayOfMonth(), 23, 0);
                            Log.d("dateTime", taskDateTime.toString());
                            taskDateTime = taskDateTime.withTime(localTime);
                            timeText.setText(taskDateTime.toString("h:mm a"));


                        } else {
                            taskDateTime = taskDateTime.withTime(localTime);
                            timeText.setText(taskDateTime.toString("MMM d, h:mm a"));

                        }

                        if (editing) {
                            if (taskDateTime.isAfter(DateTime.now()))
                                notified = false;
                            else if (taskDateTime.isBefore(taskList.get(taskPosition).getDateTime()))
                                notified = taskList.get(taskPosition).isNotified();
                        }

                        Log.d("dateTime", taskDateTime.toString("h:mm:ss"));

                        resetTimeButton.setVisibility(View.VISIBLE);


                    }
                }, DateTime.now().getHourOfDay(), DateTime.now().getMinuteOfHour(), false);
                timePickerDialog.show();

/*                TimePickerFragmentDialog.newInstance(DateTimeBuilder.newInstance()
                        .withMinDate(timeMin))
                        .show(getSupportFragmentManager(), "TimePickerFragmentDialog");*/
            }
        });

        timeText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                taskDateTime = new DateTime(3000, 1, 1, 0, 0);
                timeText.setText("Add time");
            }
        });

    }


    public void fabAddList() {
        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                checkFull();
                if (full) {
                    saveTask();
                    finish();
                } else {
                    Flashbar flashbar = blankText();
                    flashbar.show();
                }
            }
        });
    }


    public void saveTask() {
        EditText taskName = (EditText) findViewById(R.id.taskName);
        ArrayList<String> items = new ArrayList<>();
        ArrayList<Boolean> checked = new ArrayList<>();
        ArrayList<Integer> tags = new ArrayList<>(AddTaskTagAdapter.tagList);
        Switch switch_task = (Switch) findViewById(R.id.switch_task);
        if (switch_task.isChecked()) {
            Log.d("Adding Items", "Adding Items");
            items = new ArrayList<>(itemsList);
        } else {
            //AddTaskAdapter.taskList.clear();
        }


        if (editing) {
            TaskListInterface.replaceTask(taskName.getText().toString(), items, checked, tags, taskDateTime, notified, TaskList.taskList.indexOf(taskList.get(taskPosition)), taskList.get(taskPosition).getListKey(), getApplicationContext());
        } else {
            TaskListInterface.addTask(taskName.getText().toString(), items, checked, tags, taskDateTime);
        }

    }


    public static void addTag(int tagPosition) {
        boolean isAlreadyAdded = false;
        for (int i = 0; i < AddTaskTagAdapter.tagList.size(); i++) {
            if (AddTaskTagAdapter.tagList.get(i) == tagPosition) {
                isAlreadyAdded = true;
            }
        }
        if (!isAlreadyAdded) {
            AddTaskTagAdapter.tagList.add(tagPosition);
            Log.d("addingValue", "adding #" + tagPosition);
            tagAdapter.notifyDataSetChanged();
        } else {
            Flashbar flashbar = tagExists(activity);
            flashbar.show();
        }
    }


    public void checkFull() {
        Log.d("taskName", "gets here");

        EditText taskName = (EditText) findViewById(R.id.taskName);
        Switch switch_task = (Switch) findViewById(R.id.switch_task);
        if (TextUtils.isEmpty(taskName.getText())) {
            full = false;
            Log.d("taskName", "full = false");
        } else if (switch_task.isChecked()) {


            for (int i = 0; i < itemsList.size(); i++) {
                if (itemsList.get(i).isEmpty()) {
                    full = false;
                    break;
                } else {
                    full = true;
                }
            }
        } else if (taskDateTime == null) {

        } else {
            full = true;
        }
    }

    private Flashbar blankText() {
        return new Flashbar.Builder(this)
                .gravity(Flashbar.Gravity.BOTTOM)
                .title("Blank field")
                .message("Please fill in or remove any blank fields")
                .dismissOnTapOutside()
                .backgroundColor(Cyanea.getInstance().getAccent())
                .build();

    }

    private static Flashbar tagExists(Activity activity) {
        return new Flashbar.Builder(activity)
                .gravity(Flashbar.Gravity.BOTTOM)
                .title("Tag Exists")
                .message("The tag you selected has already been added to this task")
                .dismissOnTapOutside()
                .backgroundColor(Cyanea.getInstance().getAccent())
                .build();

    }


    @Override
    public void onDateSet(long date) {





    }

    @Override
    public void onTimeSet(long time, long date) {

    }


    // Extend the Callback class
    ItemTouchHelper.Callback _ithCallback = new ItemTouchHelper.Callback() {
        //and in your implementation of
        public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {

            int fromPosition = viewHolder.getAdapterPosition();
            int toPosition = target.getAdapterPosition();

            if (fromPosition < toPosition) {
                for (int i = fromPosition; i < toPosition; i++) {
                    Log.d("dragDropStart", itemsList.get(fromPosition));
                    Log.d("dragDropStart", itemsList.get(toPosition));
                    Collections.swap(itemsList, i, i + 1);
                    Log.d("dragDropEnd", itemsList.get(fromPosition));
                    Log.d("dragDropEnd", itemsList.get(toPosition));
                }
            } else {
                for (int i = fromPosition; i > toPosition; i--) {
                    Log.d("dragDropStart", itemsList.get(fromPosition));
                    Log.d("dragDropStart", itemsList.get(toPosition));
                    Collections.swap(itemsList, i, i - 1);
                    Log.d("dragDropEnd", itemsList.get(fromPosition));
                    Log.d("dragDropEnd", itemsList.get(toPosition));
                }
            }
            Log.d("dragDropStart", itemsList.get(fromPosition));
            Log.d("dragDropStart", itemsList.get(toPosition));
            mAdapter.notifyItemMoved(fromPosition, toPosition);
            Log.d("dragDropEnd", itemsList.get(fromPosition));
            Log.d("dragDropEnd", itemsList.get(toPosition));

            return false;


        }

        @Override
        public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
        }

        @Override
        public boolean isLongPressDragEnabled() {

            return true;
        }

        @Override
        public boolean isItemViewSwipeEnabled() {
            return false;
        }

        //defines the enabled move directions in each state (idle, swiping, dragging).
        @Override
        public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
            return makeFlag(ItemTouchHelper.ACTION_STATE_DRAG,
                    ItemTouchHelper.DOWN | ItemTouchHelper.UP);
        }
    };


    @Override
    public void onPause() {
        super.onPause();
    }


}

File: app/src/main/java/com/andb/apps/todo/AddTask.kt
package com.andb.apps.todo

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.PorterDuff
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.andb.apps.todo.eventbus.AddTaskAddTagEvent
import com.andb.apps.todo.notifications.NotificationHandler
import com.andb.apps.todo.objects.Tasks
import com.andb.apps.todo.utilities.Current
import com.andb.apps.todo.utilities.ProjectsUtils
import com.andb.apps.todo.utilities.Utilities
import com.andb.apps.todo.views.ReminderPicker
import com.github.rongi.klaster.Klaster
import com.google.android.material.snackbar.Snackbar
import com.jaredrummler.cyanea.Cyanea
import kotlinx.android.synthetic.main.add_task.view.*
import kotlinx.android.synthetic.main.add_task_list_item.view.*
import kotlinx.android.synthetic.main.task_view_tag_list_item.view.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.*
import kotlin.collections.ArrayList


class AddTask(ctxt: Context, val activity: MainActivity) : FrameLayout(ctxt) {


    private lateinit var subtaskAdapter: RecyclerView.Adapter<RecyclerView.ViewHolder>
    private lateinit var tagAdapter: RecyclerView.Adapter<RecyclerView.ViewHolder>

    private lateinit var task: Tasks

    private var dateSet = false

    init {
        inflate(context, R.layout.add_task, this)
        EventBus.getDefault().register(this)
    }

    fun setup(task: Tasks) {
        this.task = task
        theme()
        setupName()
        setupRecyclers()
        setupButtons()
        setReminders()
    }


    private fun theme() {
        addTaskEditFrame.setBackgroundColor(Utilities.lighterDarker(Cyanea.instance.backgroundColor, .95f))
        addTaskName.apply {
            setHintTextColor(Utilities.sidedLighterDarker(Cyanea.instance.backgroundColor, .7f))
            setTextColor(Utilities.sidedLighterDarker(Cyanea.instance.backgroundColor, .5f))
            setCursorColor(this, Utilities.sidedLighterDarker(Cyanea.instance.backgroundColor, .6f))
        }
        checkDividers()
    }


    private fun setupName() {
        addTaskName.apply {
            setText(task.listName)
            addTextChangedListener(object : TextWatcher {
                override fun afterTextChanged(s: Editable?) {}
                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                    task.listName = s.toString()
                }
            })
        }
    }


    private fun setupRecyclers() {
        val tagRecyclerView = addTaskTagsRV
        tagRecyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        tagAdapter = tagAdapter()
        tagRecyclerView.adapter = tagAdapter

        val taskRecycler = addTaskSublistRV
        taskRecycler.layoutManager = LinearLayoutManager(context)
        subtaskAdapter = subtaskAdapter()
        taskRecycler.adapter = subtaskAdapter
        val ith = ItemTouchHelper(_ithCallback)
        ith.attachToRecyclerView(taskRecycler)
    }

    private fun setupButtons() {
        addTaskAddListIcon.setOnClickListener {
            addSubtask()
            checkDividers()
        }
        addTaskAddTagsIcon.setOnClickListener {
            addTag()
            checkDividers()
        }
        addTaskAddReminderIcon.setOnClickListener {
            addReminder()
        }
        addTaskConfirm.setOnClickListener {
            save()
        }
    }

    @SuppressLint("SetTextI18n")
    private fun setReminders() {
        addTaskRemindersText.text = (task.timeReminders.size + task.locationReminders.size).toString() + " " + context.getString(R.string.reminder_picker_title)
    }

    private fun addReminder() {
        val view = ReminderPicker(context)
        val dialog = AlertDialog.Builder(context).setView(view).show().also {
            it.window?.setBackgroundDrawable(null)
            it.setOnCancelListener {
                setReminders()
            }
        }
        view.setup(task, dialog)
    }

    private fun checkDividers() {
        addTaskDivider1.visibility = if (task.isListItems) {
            View.VISIBLE
        } else {
            View.GONE
        }
        addTaskDivider2.visibility = if (task.isListTags) {
            View.VISIBLE
        } else {
            View.GONE
        }
    }

    private fun addSubtask() {
        task.listItems.add("")
        task.listItemsChecked.add(false)
        subtaskAdapter.notifyItemInserted(task.listItems.size - 1)
    }

    private fun addTag() {
        val selectTag = Intent(context, TagSelect::class.java)
        selectTag.putExtra("isTaskCreate", true)
        context.startActivity(selectTag)
    }


    private fun validate(): Boolean {
        if (task.listName.isEmpty()) {
            Snackbar.make(addTaskEditFrame.rootView, "Please provide a name for the task", Snackbar.LENGTH_LONG)
                .setAnimationMode(Snackbar.ANIMATION_MODE_SLIDE).show()
            return false
        }
        when (task.listName) {
            "OVERDUE", "TODAY", "WEEK", "MONTH", "FUTURE", "ADD_TASK_PLACEHOLDER" -> {
                Snackbar.make(addTaskEditFrame.rootView, "Sorry, those names are reserved for the app's use. Please choose a different wording or capitalization", Snackbar.LENGTH_LONG)
                    .setAnimationMode(Snackbar.ANIMATION_MODE_SLIDE).show()
                return false
            }
        }

        val tempList: ArrayList<Int> = ArrayList()
        for ((i, l) in task.listItems.withIndex()) {
            if (l.isEmpty()) {
                tempList.add(i)
            }
        }
        for (i in tempList.size - 1 downTo 0) {
            task.listItems.removeAt(tempList[i])
        }

        return true
    }

    fun save() {

        if (validate()) {
            update()
            ProjectsUtils.update(task)
            NotificationHandler.resetNotifications()
        }
    }

    private fun update() {
        task.isEditing = false
        activity.inboxFragment.adding = false
        activity.inboxFragment.editingId = -1
    }


    private fun tagAdapter() = Klaster.get()
        .itemCount { task.listTagsSize }
        .view(R.layout.task_view_tag_list_item, LayoutInflater.from(context))
        .bind { _ ->
            val tag = Current.tagListAll()[task.listTags[adapterPosition]]
            itemView.apply {
                tagImage.setColorFilter(tag.tagColor)
                task_view_item_tag_name.text = tag.tagName
                setOnClickListener {
                    task.listTags.removeAt(adapterPosition)
                    tagAdapter.notifyItemRemoved(adapterPosition)
                    checkDividers()
                }
            }

        }
        .build()

    private fun subtaskAdapter(): RecyclerView.Adapter<RecyclerView.ViewHolder> = Klaster.get()
        .itemCount { task.listItems.size }
        .view(R.layout.add_task_list_item, LayoutInflater.from(context))
        .bind { _ ->
            itemView.apply {
                taskItemEditText.apply {
                    setText(task.listItems[adapterPosition])
                    addTextChangedListener(object : TextWatcher {
                        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                            task.listItems[adapterPosition] = s.toString()
                        }

                        override fun afterTextChanged(s: Editable?) {}
                        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
                    })
                    backgroundTintList = ColorStateList.valueOf(Utilities.lighterDarker(Cyanea.instance.backgroundColor, .6f))
                    setTextColor(Utilities.lighterDarker(Cyanea.instance.backgroundColor, .7f))
                    setCursorColor(this, Utilities.lighterDarker(Cyanea.instance.backgroundColor, .6f))
                }
                removeListItem.setOnClickListener {
                    task.listItems.removeAt(adapterPosition)
                    task.listItemsChecked.removeAt(adapterPosition)
                    subtaskAdapter.notifyItemRemoved(adapterPosition)
                    checkDividers()
                }
                if (adapterPosition == task.listItems.size - 1) {
                    (layoutParams as RecyclerView.LayoutParams).bottomMargin = Utilities.pxFromDp(8)
                } else {
                    (layoutParams as RecyclerView.LayoutParams).bottomMargin = Utilities.pxFromDp(0)
                }
            }
        }
        .build()


    private fun setCursorColor(editText: EditText, color: Int) {
        try {
            // Get the cursor resource id
            var field = TextView::class.java.getDeclaredField("mCursorDrawableRes")
            field.isAccessible = true
            val drawableResId = field.getInt(editText)

            // Get the drawable and set a color filter
            val drawable = ContextCompat.getDrawable(editText.context, drawableResId)
            drawable?.setColorFilter(color, PorterDuff.Mode.SRC_IN)
            val drawables = arrayOf(drawable, drawable)


            // Get the editor
            field = TextView::class.java.getDeclaredField("mEditor")
            field.isAccessible = true
            val editor = field.get(editText)
            // Set the drawables
            field = editor.javaClass.getDeclaredField("mCursorDrawable")
            field.isAccessible = true
            field.set(editor, drawables)

        } catch (e: Exception) {
            Log.e("cursorSetFailed", "-> ", e)
        }
    }

    private var _ithCallback: ItemTouchHelper.Callback = object : ItemTouchHelper.Callback() {
        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            val fromPosition = viewHolder.adapterPosition
            val toPosition = target.adapterPosition

            if (fromPosition < toPosition) {
                for (i in fromPosition until toPosition) {
                    Collections.swap(task.listItems, i, i + 1)
                    Collections.swap(task.listItemsChecked, i, i + 1)
                }
            } else {
                for (i in fromPosition downTo toPosition + 1) {
                    Collections.swap(task.listItems, i, i - 1)
                    Collections.swap(task.listItemsChecked, i, i - 1)
                }
            }
            subtaskAdapter.notifyItemMoved(fromPosition, toPosition)

            return false
        }

        //defines the enabled move directions in each state (idle, swiping, dragging).
        override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
            return ItemTouchHelper.Callback.makeFlag(
                ItemTouchHelper.ACTION_STATE_DRAG,
                ItemTouchHelper.DOWN or ItemTouchHelper.UP
            )
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {}
        override fun isLongPressDragEnabled(): Boolean {
            return true
        }

        override fun isItemViewSwipeEnabled(): Boolean {
            return false
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onAddTaskAddTagEvent(event: AddTaskAddTagEvent) {
        if (!task.listTags.contains(event.tag)) {
            task.listTags.add(event.tag)
            tagAdapter.notifyItemInserted(task.listTagsSize - 1)
            checkDividers()
        } else {
            Snackbar.make(addTaskEditFrame.rootView, "The tag you selected has already been added to this task", Snackbar.LENGTH_LONG)
                .setAnimationMode(Snackbar.ANIMATION_MODE_SLIDE).show()
        }
    }


}








